Layer panel v1

kfarr opened the pull request:
This is an attempted redo of https://github.com/3DStreet/3dstreet-editor/pull/3 broken into smaller commits

kfarr committed:
new icons for layer panel

kfarr committed:
first pass at new styles

still to fix:
- arrow not clickable
- checkboxes don't yet follow indentation per design (just replaced eyeballs for v1)
- still need to replace entity tag with custom data name
- a few more optimizations

kfarr commented:
current state

![image](https://user-images.githubusercontent.com/470477/176973980-683881f4-6a12-4c4f-a018-2657596c6c54.png)

kfarr committed:
properly indent layers and show/hide children

- indentations now work for layers as designed; checkbox placed correctly
- arrows to show / hide children now work as expected
- added workaround for safari, the `entityCloseTag` class in src//lib/entity.js seems to be needed to make safari draw the full width of the left panel

kfarr committed:
custom display names and safari breaks

- use `data-layer-name` for custom entity names if available, fallback to class, mixin, or tag
- this breaks safari - the left-hand panel is half-width and doesn't show dropdown icon

kfarr commented:
Current state in chrome
![image](https://user-images.githubusercontent.com/470477/176987077-9f80f754-9e45-43d4-82c2-bb6851148238.png)

kfarr committed:
left menu toggle

kfarr commented:
current state
miraculously works in safari (right) again
![image](https://user-images.githubusercontent.com/470477/176987682-d1a47ef3-2de5-4514-9380-90f958b7e483.png)

kfarr committed:
only show entities with data-layer-name

kfarr committed:
wrap top level entities in div with layer class

kfarr commented:
@Algorush can you work from this draft PR and address outstanding issues? cc @srothst1

1. padding on div entity option (see doc linked below)
2. hover state for expand / collapse (see doc linked below)
3. safari does not display entire left panel
4. match font (work sans) and font weight to figma

See notes on items 1 and 2 here:
https://docs.google.com/document/d/1nIx43M2ZLGtELJNmCHqlmjTDDRIZ10ohQsDP82t5QzU/edit

srothst1 commented:
One quick comment / question before I submit a code review:
- Should the items in each layer have a drop down arrow. For instance, should 16th St West of Harrison have a selectable drop down arrow?
![Screen Shot 2022-07-03 at 5 35 34 PM](https://user-images.githubusercontent.com/39531367/177063202-fe08fb63-42a4-432d-8b41-67c4a40b7c4c.png)
This issue appears to be documented [here](https://github.com/3DStreet/3dstreet-editor/pull/4/commits/8978fa49f6ed05a767ecfc297c5d1b3086b291be) 👍 Feel free to hide / close this comment if appropriate.

srothst1 commented on src/lib/entity.js:
```diff
@@ -556,15 +556,19 @@ export function printEntity(entity, onDoubleClick) {
-
+  // Custom display name for a layer if available, otherwise use entity name or tag
+  let displayName = entity.getAttribute('data-layer-name');
+  if (!displayName) {
+    displayName = entityName;
+  }
+  if (!displayName) {
+    displayName = entity.tagName.toLowerCase();
+  }
```
I don't understand why there are two if statements here. Is there any reason why _just_ this wouldn't be a cleaner solution:
```
  if (!displayName) {
    displayName = entity.tagName.toLowerCase();
  }
```

Algorush replied:
> Should the items in each layer have a drop down arrow. For instance, should 16th St West of Harrison have a selectable drop down arrow?

Hi, @srothst1. Yes, I am aware of this issue. I wanted to ask about the best way to fix it. The arrows appear under the existing aframe-inspector code. Because it checks if the entity has elements in the children array. Here is the code in "components/scenegraph/Entity.js":
```
    let collapse;
    if (entity.children.length > 0 &&  
        !isFiltering) {
      collapse = (
        <span
...
```
The point is that entities with data-layer-name='16th St West of Harrison' and data-layer-name='16th St East of Harrison' have a link to the parent entity with id='streets' in their children array. I don't know yet where this mechanism is implemented. For entities with data-layer-name='3D Scan' and 'Map', this does not happen

kfarr replied:
> I don't understand why there are two if statements here. Is there any reason why just this wouldn't be a cleaner solution:
> 
>   if (!displayName) {
>     displayName = entity.tagName.toLowerCase();
>   }

Hey @srothst1 sorry that is confusing. entityName is not guaranteed, so this part you quoted is a backup to show the tag if nothing else if available to describe the entity. displayName cascades from data-layer-name to id, class, mixin, or if none of those are found, the tag converted to lowercase. You can see it in action on this screenshot where cameraRig is an id, a-scene is a tag, Street Segments is from data-layer-name: https://github.com/3DStreet/3dstreet-editor/pull/4#issuecomment-1172834844

This could be useful if we want to have a "dev mode" switch where we show all entities and properties.

the logic is perhaps ugly but it works, similar to the endless "else if" statements in aframe-streetmix-parsers.js

kfarr replied:
> > Should the items in each layer have a drop down arrow. For instance, should 16th St West of Harrison have a selectable drop down arrow?

> The point is that entities with data-layer-name='16th St West of Harrison' and data-layer-name='16th St East of Harrison' have a link to the parent entity with id='streets' in their children array. I don't know yet where this mechanism is implemented. For entities with data-layer-name='3D Scan' and 'Map', this does not happen

The logic appears to be here:
https://github.com/3DStreet/3dstreet-editor/blob/master/src/components/scenegraph/Entity.js#L67
`    if (entity.children.length > 0 && !isFiltering) {`

Perhaps we could add a check for `data-layer-show-children` property on the entity. The new line 67 could be:
`    if (entity.children.length > 0 && !isFiltering && !!entity.getAttribute('data-layer-show-children')) {`

and the entities in index.html for Imagery and Street Segments will need `data-layer-show-children="true"`

srothst1 replied:
Hi everyone. Seems like we are having two conversations at once in this thread. To be clear:

1. Regarding [this comment](https://github.com/3DStreet/3dstreet-editor/pull/4#discussion_r912575092) about the `Entity` arrow, I like @kfarr's [approach](https://github.com/3DStreet/3dstreet-editor/pull/4#discussion_r912606471). However, it might be simpler to use an `<a-entity>` property that already exists in `index.html`. The following code in ` ` seems to work just find for now:
```js
let collapse;
if (entity.children.length > 0 && !isFiltering && !entity.hasAttribute('position')) {
  collapse = (
    <span
      onClick={() => this.props.toggleExpandedCollapsed(entity)}
      className={`collapsespace fa ${
        isExpanded ? 'fa-caret-down' : 'fa-caret-up'
      }`}
    />
  );
} else {
  collapse = <span className="collapsespace" />;
}
``` 
This does, of course, assume that all streets added to the scene will have the attribute `position`. I feel like this is a safe assumption. @kfarr and @Algorush please let me know what you think about this code. If you both approve, I will commit the change to this pull request.

2. Regarding my [original question](https://github.com/3DStreet/3dstreet-editor/pull/4#discussion_r912567259), I really appreciate @kfarr's [explanation](https://github.com/3DStreet/3dstreet-editor/pull/4#discussion_r912600479). If `entityName` is `''` in certain circumstances. We may be able to use a ternary statement to simplify the code:
```js
if (!displayName) {
  displayName = entityName === '' ? entity.tagName.toLowerCase() : entityName;
}
```
Still not 100% sure I understand the current code. However, if it is currently working we may not want to overthink it.

kfarr replied:
Re #1 Thanks for the idea about using position tag, it is clever but I would prefer being more explicit about the behavior, so I went ahead and updated the code to check `hasAttribute('data-layer-show-children')`

Re #2 entityName might be null, empty or other weird javascript falsy values besides empty string so I'm going to keep the current code just in case since it works

srothst1 commented on src/components/scenegraph/SceneGraph.js:
```diff
@@ -247,26 +248,52 @@ export default class SceneGraph extends React.Component {
+              isExpanded={this.isExpanded(entityOption.entity)}
+              isSelected={this.props.selectedEntity === entityOption.entity}
+              selectEntity={this.selectEntity}
+              toggleExpandedCollapsed={this.toggleExpandedCollapsed}
```
Is it possible to remove the arrow to toggle expand / collapsed list? cc: https://github.com/3DStreet/3dstreet-editor/pull/4#issuecomment-1173212414

kfarr replied:
addressed with https://github.com/3DStreet/3dstreet-editor/pull/4/commits/ae2c6711726fda296b8626efeed6e16e34e80ae3

srothst1 reviewed (commented):
Here is my initial review. I did not look at the `CSS` portion of this pull request.

kfarr committed:
show layer expand toggle only if entity has `data-layer-show-children`

kfarr commented:
@Algorush we are still awaiting updates on your edits on items 1, 2, 3 and 4 below:

- [x] 1. padding on div entity option (see notes in [this doc](https://docs.google.com/document/d/1nIx43M2ZLGtELJNmCHqlmjTDDRIZ10ohQsDP82t5QzU/edit))
- [x] 2. hover state for expand / collapse (see notes in [this doc](https://docs.google.com/document/d/1nIx43M2ZLGtELJNmCHqlmjTDDRIZ10ohQsDP82t5QzU/edit))
- [ ] 3. safari does not display entire left panel
- [x] 4. match font (work sans) and font weight to figma

Algorush committed:
Improve left-panel styles

add new code and styles for selected entities
add hover and pressed to toggle-leftbar

Algorush committed:
Merge branch 'layer-panel-v1' of https://github.com/3DStreet/3dstreet-editor into layer-panel-v1

Algorush commented:
Hi all! I did points 1 and 2. Please check. Added new styles for selected entities. To do this, I had to add a little code so that the design matches Figma.

kfarr commented:
@Algorush great! On point 4 it's almost there with your recent commit, last part is the Layers title needs work sans as well
![image](https://user-images.githubusercontent.com/470477/177471774-d6e4c344-c185-4442-8120-ba9af4deb4d2.png)

Algorush committed:
add prevent entity select when clicking on arrow or checkbox

add changing camera dopdown indicator

Algorush commented:
I did everything except Safari support. Check, please. I haven't been able to run Safari on my Win 10 yet

kfarr committed:
fix default scene to 16th and svn

- fix alignment with mapbox view
- reposition streets and 3dtiles entities
- update title

kfarr commented:
Everything works except safari support. I think we should merge this and start a safari fix PR.
